llm/vicuna-llama-2/scripts/train.py

```diff
@@ -0,0 +1,346 @@
+        self.tokenizer = tokenizer
+
+        rank0_print("Formatting inputs...Skip in lazy mode")
+        self.tokenizer = tokenizer
+        self.raw_data = raw_data
+        self.cached_data_dict = {}
+
+    def __len__(self):
+        return len(self.raw_data)
+
+    def __getitem__(self, i) -> Dict[str, torch.Tensor]:
+        if i in self.cached_data_dict:
+            return self.cached_data_dict[i]
+
+        ret = preprocess([self.raw_data[i]["conversations"]], self.tokenizer)
+        ret = dict(
+            input_ids=ret["input_ids"][0],
+            labels=ret["labels"][0],
+            attention_mask=ret["attention_mask"][0],
+        )
+        self.cached_data_dict[i] = ret
+
+        return ret
+
+
+def make_supervised_data_module(tokenizer: transformers.PreTrainedTokenizer,
+                                data_args) -> Dict:
+    """Make dataset and collator for supervised fine-tuning."""
+    dataset_cls = (LazySupervisedDataset
+                   if data_args.lazy_preprocess else SupervisedDataset)
+    rank0_print("Loading data...")
+
+    train_json = json.load(open(data_args.data_path, "r"))
+    train_dataset = dataset_cls(train_json, tokenizer=tokenizer)
+
+    if data_args.eval_data_path:
+        eval_json = json.load(open(data_args.eval_data_path, "r"))
+        eval_dataset = dataset_cls(eval_json, tokenizer=tokenizer)
+    else:
+        eval_dataset = None
+
+    return dict(train_dataset=train_dataset, eval_dataset=eval_dataset)
+
+
+class CheckpointCallback(transformers.TrainerCallback):
+
+    def on_save(self, args, state, control, **kwargs):
+        """Add complete indicator to avoid incomplete checkpoints."""
+        if state.is_world_process_zero:
+            ckpt_path = os.path.join(args.output_dir,
+                                     f'checkpoint-{state.global_step}')
+            with open(os.path.join(ckpt_path, 'complete'), 'w') as f:
+                f.write('')
+            print(f'Checkpoint {state.global_step} saved.')
+        torch.distributed.barrier()
+
+
+def cleanup_incomplete_checkpoints(output_dir):
+    """Remove incomplete checkpoints."""
+    checkpoints = list(pathlib.Path(output_dir).glob('checkpoint-*'))
+    checkpoints = [c for c in checkpoints if c.name.split('-')[-1].isdigit()]
+    checkpoints = sorted(checkpoints,
+                         key=lambda x: int(x.name.split('-')[-1]),
+                         reverse=True)
+    for checkpoint in checkpoints:
+        if not (checkpoint / 'complete').exists():
+            print(f'Removing incomplete checkpoint {checkpoint}')
+            shutil.rmtree(checkpoint)
+        else:
+            print(f'Using checkpoint {checkpoint}, copying to ~/tmp/ for '
+                  'optimization of loading.')
+            tmp_dir = os.path.expanduser('~/tmp')
+            os.makedirs(tmp_dir, exist_ok=True)
+            subprocess.run(['gsutil', '-m', 'rsync', '-r', checkpoint, tmp_dir],
```

Comment:
Worth a comment there: Forcing the mount tool to download the files first to warm the cache. This significantly speeds up checkpoint loading.